tests/testScrapPaper.js

```diff
@@ -0,0 +1,386 @@
+  config.authName = user.authName;
+  config.authUin = user.authUin;
+};
+
+/**
+ * Gets an instance question URL for the current user
+ * @param {string} baseUrl
+ * @param {string} hm1AutomaticTestSuiteUrl
+ * @returns {string}
+ */
+const getBarcodeSubmissionUrl = async (baseUrl, hm1AutomaticTestSuiteUrl) => {
+  const res = await fetch(hm1AutomaticTestSuiteUrl);
+  assert.equal(res.ok, true);
+
+  const hm1Body = await res.text();
+  assert.include(hm1Body, 'HW1.12. Barcode submission');
+
+  const $hm1Body = cheerio.load(hm1Body);
+  return (
+    baseUrl.replace('/pl', '') + $hm1Body('a:contains("HW1.12. Barcode submission")').attr('href')
+  );
+};
+
+const getScrapPaperPayload = ($page, numPages, pageLabel) => {
+  return {
+    num_pages: numPages,
+    page_label: pageLabel,
+    __csrf_token: $page('form > input[name="__csrf_token"]').val(),
+    __action: $page('form > input[name="__action"]').val(),
+  };
+};
+
+// const getScanPaperPayload = ($page, pdfBuffer) => {
+//   const formData = new FormData();
+//   formData.append('file', pdfBuffer, 'ANY FILENAME.pdf');
+//   formData.append('__csrf_token', $page('form > input[name="__csrf_token"]').val());
+//   formData.append('__action', $page('form > input[name="__action"]').val());
+//   return formData;
+// };
+
+describe('Barcode generation, student submission, and scanning process', function () {
+  this.timeout(75000);
+  const baseUrl = 'http://localhost:' + config.serverPort + '/pl';
+  const scrapPaperUrl = baseUrl + '/scrap_paper';
+  // const scanPaperUrl = baseUrl + '/scan_paper';
+  const mockStudents = [
+    { authUid: 'student1', authName: 'Student User 1', authUin: '00000001' },
+    { authUid: 'student2', authName: 'Student User 2', authUin: '00000002' },
+  ];
+
+  // created in `Generate scrap paper` but also used in `Barcode submission ..` and `Scan scrap paper` test blocks in end-to-end test
+  let barcodeRows;
+  let pdfBuffer;
+  let hm1AutomaticTestSuiteUrl;
+  let defaultUser;
+
+  before('set up testing server', async () => {
+    await util.promisify(helperServer.before().bind(this))();
+  });
+  after('shut down testing server', helperServer.after);
+
+  describe('Generate scrap paper', function () {
+    let $scrapPaper;
+
+    describe('GET', function () {
+      before('fetch page', async () => {
+        const res = await fetch(scrapPaperUrl);
+        assert.equal(res.status, 200);
+        $scrapPaper = cheerio.load(await res.text());
+      });
+
+      it('view should display PDF generation form', () => {
+        const numPages = $scrapPaper('#num_pages');
+        const pageLabel = $scrapPaper('#page_label');
+        assert.lengthOf(numPages, 1);
+        assert.lengthOf(pageLabel, 1);
+      });
+    });
+
+    describe('POST', () => {
+      let pdf;
+      // let decodedJpegs;
+
+      it('user should be able to download a pdf', async () => {
+        const payload = getScrapPaperPayload($scrapPaper, testNumPages, testLabel);
```

Comment:
AFAICT this is the only place `getScrapPaperPayload` is used. Let's inline it so that we can more easily see what this test is doing.

Reply:
Done.